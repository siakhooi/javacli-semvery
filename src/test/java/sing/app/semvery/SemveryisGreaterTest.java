package sing.app.semvery;

import static com.github.stefanbirkner.systemlambda.SystemLambda.tapSystemOut;
import static org.junit.jupiter.api.Assertions.assertDoesNotThrow;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static sing.app.semvery.TestData.createArguments;
import java.util.ArrayList;
import java.util.stream.Stream;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.extension.ExtendWith;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.Arguments;
import org.junit.jupiter.params.provider.MethodSource;
import au.com.origin.snapshots.Expect;
import au.com.origin.snapshots.junit5.SnapshotExtension;

@ExtendWith({SnapshotExtension.class})
class SemveryisGreaterTest {

    private Semvery app;
    private Expect expect;

    @BeforeEach
    void setup() {
        this.app = new Semvery();
    }

    @ParameterizedTest
    @MethodSource
    void testIsGreater(String scenario, ReturnValue returnValue, String[] arguments)
            throws Exception {
        String text = tapSystemOut(() -> assertEquals(returnValue, app.run(arguments)));
        assertDoesNotThrow(() -> expect.scenario(scenario).toMatchSnapshot(text));
    }

    private static Stream<Arguments> testIsGreater() {
        ArrayList<Arguments> a = new ArrayList<>();

        final ReturnValue ok = ReturnValue.OK;
        final ReturnValue nok = ReturnValue.NOT_OK;
        final String OPERATOR = "isGreater";
        final String[] DEFAULT = new String[] {"-r", "2.0.0"};
        final String[] ANY = new String[] {"-r", "2.0.0", "--any"};

        for (String o : TestData.OPERATIONS) {
            a.addAll(createArguments("all-greater-2.0.0", ok, o, OPERATOR, DEFAULT,
                    TestData.ALL_GREATER_2_0_0));
            a.addAll(createArguments("all-greater-2.0.0", ok, o, OPERATOR, ANY,
                    TestData.ALL_GREATER_2_0_0));

            a.addAll(createArguments("all-not-greater-2.0.0", nok, o, OPERATOR, DEFAULT,
                    TestData.ALL_NOT_GREATER_2_0_0));
            a.addAll(createArguments("all-not-greater-2.0.0", nok, o, OPERATOR, ANY,
                    TestData.ALL_NOT_GREATER_2_0_0));

            a.addAll(createArguments("all-not-valid", nok, o, OPERATOR, DEFAULT,
                    TestData.ALL_NOT_VALID));
            a.addAll(createArguments("all-not-valid", nok, o, OPERATOR, ANY, TestData.ALL_NOT_VALID));

            a.addAll(createArguments("mixed-greater-not-greater-2.0.0", nok, o, OPERATOR, DEFAULT,
                    TestData.GREATER_NOT_GREATER_2_0_0));
            a.addAll(createArguments("mixed-greater-not-greater-2.0.0", ok, o, OPERATOR, ANY,
                    TestData.GREATER_NOT_GREATER_2_0_0));

            a.addAll(createArguments("mixed-greater-2.0.0-not-valid", nok, o, OPERATOR, DEFAULT,
                    TestData.GREATER_2_0_0_NOT_VALID));
            a.addAll(createArguments("mixed-greater-2.0.0-not-valid", ok, o, OPERATOR, ANY,
                    TestData.GREATER_2_0_0_NOT_VALID));

            a.addAll(createArguments("mixed-not-greater-2.0.0-not-valid", nok, o, OPERATOR, DEFAULT,
                    TestData.NOT_GREATER_2_0_0_NOT_VALID));
            a.addAll(createArguments("mixed-not-greater-2.0.0-not-valid", nok, o, OPERATOR, ANY,
                    TestData.NOT_GREATER_2_0_0_NOT_VALID));

            a.addAll(createArguments("mixed-greater-not-greater-2.0.0-not-valid", nok, o, OPERATOR, DEFAULT,
                    TestData.GREATER_NOT_GREATER_2_0_0_NOT_VALID));
            a.addAll(createArguments("mixed-greater-not-greater-2.0.0-not-valid", ok, o, OPERATOR, ANY,
                    TestData.GREATER_NOT_GREATER_2_0_0_NOT_VALID));
        }

        return a.stream();
    }
}
