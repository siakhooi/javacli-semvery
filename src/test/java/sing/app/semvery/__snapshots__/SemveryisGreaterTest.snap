sing.app.semvery.SemveryisGreaterTest.testIsGreater[all-greater-2.0.0:2.0.1]=[
Value                Result              
-----                -----               
2.0.1                greater             

]


sing.app.semvery.SemveryisGreaterTest.testIsGreater[all-greater-2.0.0:2.0.3,4.5.7,8.7.8]=[
Value                Result              
-----                -----               
2.0.3                greater             
4.5.7                greater             
8.7.8                greater             

]


sing.app.semvery.SemveryisGreaterTest.testIsGreater[all-greater-2.0.0:2.1.0]=[
Value                Result              
-----                -----               
2.1.0                greater             

]


sing.app.semvery.SemveryisGreaterTest.testIsGreater[all-greater-2.0.0:3.1.2]=[
Value                Result              
-----                -----               
3.1.2                greater             

]


sing.app.semvery.SemveryisGreaterTest.testIsGreater[all-not-greater-2.0.0:0.1.2]=[
Value                Result              
-----                -----               
0.1.2                not greater         

]


sing.app.semvery.SemveryisGreaterTest.testIsGreater[all-not-greater-2.0.0:1.0.2]=[
Value                Result              
-----                -----               
1.0.2                not greater         

]


sing.app.semvery.SemveryisGreaterTest.testIsGreater[all-not-greater-2.0.0:1.0.3,0.5.7,2.0.0]=[
Value                Result              
-----                -----               
1.0.3                not greater         
0.5.7                not greater         
2.0.0                not greater         

]


sing.app.semvery.SemveryisGreaterTest.testIsGreater[all-not-greater-2.0.0:2.0.0]=[
Value                Result              
-----                -----               
2.0.0                not greater         

]


sing.app.semvery.SemveryisGreaterTest.testIsGreater[all-not-valid:1.0.A]=[
Value                Result              
-----                -----               
1.0.A                invalid             

]


sing.app.semvery.SemveryisGreaterTest.testIsGreater[all-not-valid:1.0.B,adfadfasjdfsd,23423,2423.2342342,24234.234234.]=[
Value                Result              
-----                -----               
1.0.B                invalid             
adfadfasjdfsd        invalid             
23423                invalid             
2423.2342342         invalid             
24234.234234.        invalid             

]


sing.app.semvery.SemveryisGreaterTest.testIsGreater[all-not-valid:23423]=[
Value                Result              
-----                -----               
23423                invalid             

]


sing.app.semvery.SemveryisGreaterTest.testIsGreater[all-not-valid:2423.2342342]=[
Value                Result              
-----                -----               
2423.2342342         invalid             

]


sing.app.semvery.SemveryisGreaterTest.testIsGreater[all-not-valid:24234.234234.]=[
Value                Result              
-----                -----               
24234.234234.        invalid             

]


sing.app.semvery.SemveryisGreaterTest.testIsGreater[all-not-valid:adfadfasjdfsd]=[
Value                Result              
-----                -----               
adfadfasjdfsd        invalid             

]


sing.app.semvery.SemveryisGreaterTest.testIsGreater[mixed-greater-2.0.0-not-valid:1.0.A,2.1.0,3.3.0]=[
Value                Result              
-----                -----               
1.0.A                invalid             
2.1.0                greater             
3.3.0                greater             

]


sing.app.semvery.SemveryisGreaterTest.testIsGreater[mixed-greater-2.0.0-not-valid:2.0.0,3.1.0,3.0.A]=[
Value                Result              
-----                -----               
2.0.0                not greater         
3.1.0                greater             
3.0.A                invalid             

]


sing.app.semvery.SemveryisGreaterTest.testIsGreater[mixed-greater-2.0.0-not-valid:2.3.0,1.1.A,3.3.0]=[
Value                Result              
-----                -----               
2.3.0                greater             
1.1.A                invalid             
3.3.0                greater             

]


sing.app.semvery.SemveryisGreaterTest.testIsGreater[mixed-greater-not-greater-2.0.0-not-valid:0.1.A,1.1.0,3.0.0]=[
Value                Result              
-----                -----               
0.1.A                invalid             
1.1.0                not greater         
3.0.0                greater             

]


sing.app.semvery.SemveryisGreaterTest.testIsGreater[mixed-greater-not-greater-2.0.0-not-valid:1.1.0,3.0.2,A.1.0]=[
Value                Result              
-----                -----               
1.1.0                not greater         
3.0.2                greater             
A.1.0                invalid             

]


sing.app.semvery.SemveryisGreaterTest.testIsGreater[mixed-greater-not-greater-2.0.0-not-valid:1.1.A,0.1.2,3.0.0]=[
Value                Result              
-----                -----               
1.1.A                invalid             
0.1.2                not greater         
3.0.0                greater             

]


sing.app.semvery.SemveryisGreaterTest.testIsGreater[mixed-greater-not-greater-2.0.0:0.1.1,1.1.0,3.0.0]=[
Value                Result              
-----                -----               
0.1.1                not greater         
1.1.0                not greater         
3.0.0                greater             

]


sing.app.semvery.SemveryisGreaterTest.testIsGreater[mixed-greater-not-greater-2.0.0:1.1.0,0.1.2,3.0.0]=[
Value                Result              
-----                -----               
1.1.0                not greater         
0.1.2                not greater         
3.0.0                greater             

]


sing.app.semvery.SemveryisGreaterTest.testIsGreater[mixed-greater-not-greater-2.0.0:1.1.0,3.0.2,0.1.0]=[
Value                Result              
-----                -----               
1.1.0                not greater         
3.0.2                greater             
0.1.0                not greater         

]


sing.app.semvery.SemveryisGreaterTest.testIsGreater[mixed-not-greater-2.0.0-not-valid:1.0.0,1.1.0,3.0.A]=[
Value                Result              
-----                -----               
1.0.0                not greater         
1.1.0                not greater         
3.0.A                invalid             

]


sing.app.semvery.SemveryisGreaterTest.testIsGreater[mixed-not-greater-2.0.0-not-valid:1.0.A,1.1.0,1.3.0]=[
Value                Result              
-----                -----               
1.0.A                invalid             
1.1.0                not greater         
1.3.0                not greater         

]


sing.app.semvery.SemveryisGreaterTest.testIsGreater[mixed-not-greater-2.0.0-not-valid:1.3.0,1.1.A,1.3.0]=[
Value                Result              
-----                -----               
1.3.0                not greater         
1.1.A                invalid             
1.3.0                not greater         

]